### ISSUE TITLE
Development using gunicorn + uvicorn

### ISSUE THREAD

**Issue by Midnighter** [Kludex/uvicorn#188]
I'm developing an API using API Star, gunicorn, and uvicorn. I would like to make my debugging life easier.

I'm able to get tracebacks about uncalled co-routines via environment variable `PYTHONASYNCIODEBUG=1` or inside the application with:

```python
asyncio.get_event_loop().set_debug(True)
```

So far so good. The asyncio docs suggest to additionally enable `ResourceWarning`s by running `python -Wdefault [...]`. Since I'm starting the application with `["gunicorn", "-c", "gunicorn.py", "app:app"]` (inside Docker), I'm unsure how to do that. I have set

```python
warnings.simplefilter("always", ResourceWarning)
```

inside my application but I don't see any warnings. Maybe I'm failing to trigger the warning even though one of my routes awaits the following co-routine:

```python
async def transport() -> str:
    LOGGER.debug("I transport.")
    trans = asyncio.Transport()
    return "Rapid"
```

According to the asyncio docs, not closing the transport should trigger the warning. I do see the logging output `'I transport.'`.

Lastly, even with extensive configuration, I'm not seeing any uvicorn log statements. I've seen this mentioned around in other issues so maybe I'm just confused whether that's solved or not. I use:

```python
logging.dictConfig({
        'version': 1,
        'disable_existing_loggers': False,
        'formatters': {
            'simple': {
                'format': "[%(name)s] [%(levelname)s] %(message)s",
            },
        },
        'handlers': {
            'console': {
                'level': 'DEBUG',
                'class': 'logging.StreamHandler',
                'formatter': 'simple'
            },
        },
        'loggers': {
            'uvicorn': {
                'level': 'DEBUG',
                'handlers': ['console']
            }
        },
        'root': {
            'level': 'DEBUG',
            'handlers': ['console'],
            'propagate': True,
        },
})
```

Adding the handler on the `uvicorn` logger and separately changing the level should not be necessary due to the changes on the `root` logger but I was kinda desperate to try something.

Any advice or help is greatly appreciated. I'm happy to share Dockerfile and minimal app if required.

The versions I'm using are:

```
Python 3.6
apistar==0.5.41
gunicorn==19.9.0
uvicorn==0.3.2
```

**Comment by Midnighter** [comment]
Hmm, I created a test file that implements some of the negative examples from the asyncio docs using 'pure' asyncio. It seems the reporting behaviour is very similar to the uvloop worker. At least on my platform running the file with `-Wdefault`, `warnings.simplefilter("always", ResourceWarning)` or neither doesn't make a difference. So I guess in the end this is just another issue about the lack of logging. The test file is:

```python
import asyncio
import logging
# import warnings


LOGGER = logging.getLogger()


async def never_called():
    LOGGER.debug("I'm never actually called.")


async def uncomsumed():
    LOGGER.debug("My exception is never consumed.")
    raise ValueError("Ouch")


async def transport():
    LOGGER.debug("Rapid")
    connect = asyncio.open_connection("http://google.com", 80)
    reader, writer = await connect


async def create():
    await asyncio.sleep(3.0)
    LOGGER.debug("(1) create file")


async def write():
    await asyncio.sleep(1.0)
    LOGGER.debug("(2) write into file")


async def close():
    LOGGER.debug("(3) close file")


async def pending():
    asyncio.ensure_future(create())
    asyncio.ensure_future(write())
    asyncio.ensure_future(close())
    await asyncio.sleep(2.0)
    loop.stop()


if __name__ == "__main__":
    # warnings.simplefilter("always", ResourceWarning)
    logging.basicConfig(level="DEBUG",
                        format="[%(name)s] [%(levelname)s] %(message)s")
    loop = asyncio.get_event_loop()
    # loop.set_debug(True)
    never_called()
    asyncio.ensure_future(uncomsumed())
    asyncio.ensure_future(transport())
    asyncio.ensure_future(pending())
    loop.run_forever()
    LOGGER.warning("Pending tasks at exit: %r.", asyncio.Task.all_tasks(loop))
    loop.close()
```

**Comment by lovelydinosaur** [comment]
Sorry, I’m slightly lost. I get the general point of “I’m not seeing resource warnings being logged”, but I’m unclear if you resolved that you still couldn’t see them in your isolated test either (ie. it’s unrelated to uvicorn) or if you could (ie. uvicorn either doesn’t log some resource errors, or doesn’t make it easy to configure them to be made visible)

**Comment by lovelydinosaur** [comment]
Probably not especially related but worth noting - uvicorn uses the root logger in 0.3.2, but uses a properly named logger in 0.3.3.

**Comment by Midnighter** [comment]
> Sorry, I’m slightly lost. I get the general point of “I’m not seeing resource warnings being logged”, but I’m unclear if you resolved that you still couldn’t see them in your isolated test either (ie. it’s unrelated to uvicorn) or if you could (ie. uvicorn either doesn’t log some resource errors, or doesn’t make it easy to configure them to be made visible)

I guess I'm also slightly confused at this point :grimacing: My point was that the asyncio documentation suggests that I should be seeing more output by tuning certain options when in reality (for my platform) this is not true. I was therefore expecting to see more information coming from uvicorn but my test case above showed that the information is the same regardless of uvicorn or asyncio.

**Comment by Midnighter** [comment]
I guess my only remaining problem is that I'm raising an exception (`ValueError`) in one of my routes and I had expected to see mention of that in the logs in debug mode. However, I only get a server error and no mention of the exception. I can, of course, catch it and log it myself but I was curious if this is expected behaviour.

**Comment by lovelydinosaur** [comment]
> I guess my only remaining problem is that I'm raising an exception (ValueError) in one of my routes and I had expected to see mention of that in the logs in debug mode. 

Indeed. That's resolved just now - version 0.3.4. Previously the debug middleware was swallowing exceptions when it returned the 500 traceback. Now we respond with the 500 traceback, but still raise the exception so that it can be caught and logged.

https://github.com/encode/uvicorn/pull/190
